File: Algorithm/88_合并两个有序数组.php
<?php
/**
 * 给定两个有序整数数组 nums1 和 nums2，将 nums2 合并到 nums1 中，使得 num1 成为一个有序数组。
 * 说明
 *     初始化 nums1 和 nums2 的元素数量分别为 m 和 n。
 *     你可以假设 nums1 有足够的空间（空间大小大于或等于 m + n）来保存 nums2 中的元素。
 * 输入 nums1 = [1,2,3,0,0,0], m = 3,nums2 = [2,5,6], n = 3
 * 输出: [1,2,2,3,5,6]
 */
class Solution
{
    /**
     * 归并排序 算法
     *     执行用时: 16 ms, 在Merge Sorted Array的PHP提交中击败了100.00% 的用户
     *     内存消耗: 14.7 MB, 在Merge Sorted Array的PHP提交中击败了100.00% 的用户
     * 时间复杂度 O(n)
     * 算法详解
     *     https://www.cnblogs.com/25-lH/p/10471966.html
     */
    function merge(&$nums1, $m, $nums2, $n)
    {
        $count = $m + $n; // 循环次数
        $n1    = $n2 = 0; // n1 n2 数组偏移量
        $arr   = [];      // 输出

        // 因为是 PHP ，所以切除数组长度之外的无用元素
        $nums1 = array_slice($nums1,0,$m);

        for ($i = 1; $i <= $count; $i++) {
            // 排除空元素的影响
            if (!isset($nums1[$n1])) {
                $arr[] = $nums2[$n2];
                $n2++;
                continue;
            }
            if (!isset($nums2[$n2])) {
                $arr[] = $nums1[$n1];
                $n1++;
                continue;
            }

            // 比较数组头
            if ($nums1[$n1] < $nums2[$n2]) {
                $arr[] = $nums1[$n1];
                $n1++;
            }
            elseif ($nums1[$n1] == $nums2[$n2]) {
                $arr[] = $nums1[$n1];
                $arr[] = $nums2[$n2];
                $n1++;
                $n2++;
                $i++;
            }
            else {
                $arr[] = $nums2[$n2];
                $n2++;
            }
        }

        $nums1 = $arr; 
    }
}

$nums1 = [-1,0,0,3,3,3,0,0,0];
$m = 6;
$nums2 = [1,2,2];
$n = 3;
$s = new Solution();
$s->merge($nums1, $m, $nums2, $n);

var_dump($nums1);

File: Go/Algorithm/88_合并两个有序数组.go
package main

import "fmt"
/**
 * 题目
 *    给定两个有序整数数组 nums1 和 nums2，将 nums2 合并到 nums1 中，使得 num1 成为一个有序数组。
 * 说明:
 *    初始化 nums1 和 nums2 的元素数量分别为 m 和 n。
 *    你可以假设 nums1 有足够的空间（空间大小大于或等于 m + n）来保存 nums2 中的元素。
 * 示例:
 *    输入:
 *        nums1 = [1,2,3,0,0,0], m = 3
 *        nums2 = [2,5,6],       n = 3
 *    输出: [1,2,2,3,5,6]
 */
func main() {

	a := []int{1,2,8,0,0,0}
	b := []int{2,5,6}

	merge(a, 3, b, 3)

	fmt.Printf("-", a)
}
/*
 * 思路
 *    归并排序
 *    归并两数组结果，赋予原数组，关于GO的底层知识运用还应该更进一步，现在还是太肤浅
 * 耗时
 *    执行用时 : 0 ms, 在Merge Sorted Array的Go提交中击败了100.00% 的用户
 *    内存消耗 : 3.9 MB, 在Merge Sorted Array的Go提交中击败了5.19% 的用户
 */
func merge(nums1 []int, m int, nums2 []int, n int)  {

	if nums2 == nil {
		return
	}

	arr := make([]int, 0)

	nPoint := 0 // nums2 指针
	mPoint := 0 // nums1 指针

	for i := 0; i <= m+n; i++ {

		if nPoint >= n && mPoint >= m  {
			continue
		}

		if nPoint >= n {
			arr = append(arr, nums1[mPoint])
			mPoint++
			continue
		}

		if mPoint >= m {
			arr = append(arr, nums2[nPoint])
			nPoint++
			continue
		}

		if nums1[mPoint] >= nums2[nPoint] {
			arr = append(arr, nums2[nPoint])
			nPoint++
		} else {
			fmt.Printf("-", mPoint)
			arr = append(arr, nums1[mPoint])
			mPoint++
		}
	}

	for i, v := range arr {
		nums1[i] = v
	}
}